File: problem304.cc
// problem304.cc

#include <iostream>
#include <vector>

using namespace std;

typedef long unsigned nat_t;

const nat_t MAX_PRIMES = 10000000;

const nat_t START = 100000000000000;
const nat_t END = 4000000;


bool is_prime(const nat_t n, const vector<nat_t>& primes) {
    for( nat_t p = 2; p*p <=n; p = primes[p] ) {
        if( n % p == 0 )
            return false;
    }
    return true;
}

int main() {
    vector<nat_t> primes(MAX_PRIMES,1);
    primes[0] = MAX_PRIMES;
    primes[1] = 0;
    primes[2] = 3;

    nat_t last_p = 2;
    for( nat_t n = 3; n < MAX_PRIMES; n += 2 ) {
        if( primes[n] == 1 ) {
            primes[last_p] = n;
            for( nat_t k = n*n; k < MAX_PRIMES; k += (2*n) ) {
                primes[k] = 0;
            }
            last_p = n;
        }
    }
    primes[last_p] = MAX_PRIMES;

    vector<bool> large(4000000, true);
    for( nat_t p = 3; p < MAX_PRIMES; p = primes[p] ) {
        nat_t n = p - START % p;
        if( n % 2 == 0 )
            n += p;
        for( ; n < END; n += (2*p) ) {
            large[n] = false;
        }
    }

    nat_t count = 0;
    for( nat_t n = 1; count < 100000; n += 2 ) {
        if( large[n] ) {
            ++count;
            cout << START + n << endl;
        }
    }
}

File: Problem304.hs
module Problem304 where

import Control.Monad
import Data.Array.ST
import Data.Array.Unboxed
import Data.List
import Data.Numbers.Primes

start, end :: Int
start = 100000000000000
end = 4000000

modulus :: Integer
modulus = 1234567891011

largePrimes :: [Int]
largePrimes = take 100000 $ [ n | (n,b) <- assocs large, b ]
    where
    large = runSTUArray $ do
                arr <- newArray (0,end) True
                forM_ (takeWhile (10000000 >) primes) $ \p -> do
                    forM_ (takeWhile (end >=) $ iterate (p +) ((-start) `mod` p)) $ \n ->
                        writeArray arr n False
                return arr

type SMat22 = (Integer,Integer,Integer)

(#*) :: SMat22 -> SMat22 -> SMat22
(a11,a12,a22) #* (b11,b12,b22) = (a11*b11+a12*b12,a11*b12+a12*b22,a12*b12+a22*b22)

(#%) :: SMat22 -> Integer -> SMat22
(a11,a12,a22) #% m = (a11 `mod` m,a12 `mod` m,a22 `mod` m)

(#^) :: SMat22 -> Int -> SMat22
(#^) = run (1,0,1)
    where
    run y x 0 = y
    run y x k = let (q,r) = k `divMod` 2
                in  run (if r == 0 then y else y #* x) (x #* x) q

multMod :: Integer -> SMat22 -> SMat22 -> SMat22
multMod m (a11,a12,a22) (b11,b12,b22) = ((a11*b11+a12*b12) `mod` m,(a11*b12+a12*b22) `mod` m,(a12*b12+a22*b22) `mod` m)

powMod :: Integer -> SMat22 -> Int -> SMat22
powMod m = run (1,0,1)
    where
    run y x 0 = y
    run y x k = let (q,r) = k `divMod` 2
                in  run (if r == 0 then y else multMod m y x) (multMod m x x) q


solution = foldl' (\s f -> let (_,a12,_) = multMod modulus f1014 f in (s+a12) `mod` modulus) 0 fs
    where f1014 = powMod modulus (0,1,1) (10^14)
          fs    = map (powMod modulus (0,1,1)) largePrimes

main = print solution

fibs = map (\(_,a12,_) -> a12) $ iterate ((0,1,1) #*) (1,0,1)